### PR TITLE
Adding pop tokens sample

### PR DESCRIPTION
**Pull request by derisen**
## Purpose

* Adds a new sample where React SPA acquiring PoP tokens to call Node.js web API which validates the token using `jose`

cc @hectormmg

## Does this introduce a breaking change?

```
[ ] Yes
[ x ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ x ] Other... Please describe: **new sample**
```

## How to Test
*  Get the code

```
git clone https://github.com/Azure-Samples/ms-identity-javascript-react-tutorial.git
cd ms-identity-javascript-react-tutorial
git checkout pop-sample
cd 6-AdvancedScenarios/2-call-api-pop/API
npm install
cd ../SPA
npm install
```

## What to Check
Verify that the following are valid
* Test in Chrome, Edge, Firefox
* Test in private mode

### CODE REVIEW

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+ 1. [Running the sample](#running-the-sample)
+ 1. [Explore the sample](#explore-the-sample)
+ 1. [About the code](#about-the-code)
+ 1. [More information](#more-information)
+ 1. [Community Help and Support](#community-help-and-support)
+ 1. [Contributing](#contributing)
+
+## Overview
+
+This sample demonstrates an React SPA calling a Node.js & Express web API that is secured using [Azure Active Directory](https://docs.microsoft.com/azure/active-directory/fundamentals/active-directory-whatis) (Azure AD). The SPA project is secured with the [Microsoft Authentication Library for React (Preview)](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-react) (MSAL React), while the web API is secured with [passport-azure-ad](https://github.com/AzureAD/passport-azure-ad).
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:18]: >an [](http://example.com/codeflow?start=25&length=2)

nit:a

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
+# React single-page application calling Node.js & Express web API protected by Azure AD using Proof of Possession
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:1]: >React single-page application calling Node.js & Express web API protected by Azure AD using Proof of Possession [](http://example.com/codeflow?start=2&length=111)

suggestion:A React single-page application calling a Node.js & Express web API using Proof of Possession

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+This sample demonstrates the [Proof of Possession](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/access-token-proof-of-possession.md) (PoP) authentication scheme. This authentication scheme cryptographically binds the access tokens to the browser and client application from which they are requested, meaning they cannot be used from a different application or device. This effectively prevents **token replay** attacks. The resource server that accepts the PoP token (i.e. a web API) needs to be able to decipher the incoming request for the PoP authentication scheme to work properly.
+
+## Scenario
+
+1. The client React SPA uses **MSAL React** to sign-in and obtain a JWT access token from **Azure AD**.
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:24]: >from [](http://example.com/codeflow?start=85&length=4)

for the Web API from..

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+1. The client React SPA wraps the access token in a secure *envelope* and stamps it with its signature.
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:25]: >The client React SPA wraps the access token in a secure *envelope* and stamps it with its signature. [](http://example.com/codeflow?start=3&length=100)

using the MSAL library??

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+1. The access token then is used with *PoP* authentication scheme to authorize the user to call the Node.js & Express web API protected by **Azure AD**.
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:26]: >Node.js & Express web API protected by **Azure AD** [](http://example.com/codeflow?start=100&length=51)

web API protected by **Azure AD**

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+1. The Node.js & Express web API decrypts the secure envelope to obtain the access token.
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:27]: >Node.js & Express [](http://example.com/codeflow?start=7&length=17)

drop

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+1. Once the access token is validated, the Node.js & Express web API responds with the protected resource.
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:28]: >the Node.js & Express web API [](http://example.com/codeflow?start=39&length=29)

just "the web API"

File: 3-Authorization-II/1-call-api/API/config.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
     "credentials": {
-        "tenantID": "Enter_the_Tenant_Info_Here",
-        "clientID": "Enter_the_Application_Id_Here"
+        "tenantID": "cbaf2168-de14-4c72-9d88-f5f05366dbef",
+        "clientID": "8c8dc348-7d2e-4c58-8c0f-15ec48bc16dc"
```
Comment (idg-sam) [inline 3-Authorization-II/1-call-api/API/config.json:4]: remove tenant and client ID

File: 3-Authorization-II/1-call-api/SPA/src/authConfig.js [hunk 9]
```diff
@@ -12,8 +12,8 @@ import { LogLevel } from "@azure/msal-browser";
  */
 export const msalConfig = {
     auth: {
-        clientId: "Enter_the_Application_Id_Here", // This is the ONLY mandatory field that you need to supply.
-        authority: "https://login.microsoftonline.com/Enter_the_Tenant_Info_Here", // Defaults to "https://login.microsoftonline.com/common"
+        clientId: "dbe9e219-febb-42b0-bbba-5f72fbc78a2e", // This is the ONLY mandatory field that you need to supply.
+        authority: "https://login.microsoftonline.com/cbaf2168-de14-4c72-9d88-f5f05366dbef", // Defaults to "https://login.microsoftonline.com/common"
```
Comment (idg-sam) [inline 3-Authorization-II/1-call-api/SPA/src/authConfig.js:16]: remove these

File: 3-Authorization-II/1-call-api/SPA/src/authConfig.js [hunk 10]
```diff
@@ -68,6 +68,6 @@ export const protectedResources = {
     },
     apiHello: {
         endpoint: "http://localhost:5000/hello",
-        scopes: ["Enter_the_Web_Api_Scope_Here"], // e.g. api://xxxxxx/access_as_user
+        scopes: ["api://8c8dc348-7d2e-4c58-8c0f-15ec48bc16dc/access_as_user"], // e.g. api://xxxxxx/access_as_user
```
Comment (idg-sam) [inline 3-Authorization-II/1-call-api/SPA/src/authConfig.js:71]: remove this

File: 6-AdvancedScenarios/2-call-api-pop/API/app.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+const config = require('./authConfig');
+const validatePoP = require('./utils/validateToken');
+const router = require('./routes/router');
+
+const BearerStrategy = require('passport-azure-ad').BearerStrategy;
+
+const options = {
+    identityMetadata: `https://${config.metadata.authority}/${config.credentials.tenantID}/${config.metadata.version}/${config.metadata.discovery}`,
+    issuer: `https://sts.windows.net/cbaf2168-de14-4c72-9d88-f5f05366dbef/`,
```
Comment (idg-sam) [inline 6-AdvancedScenarios/2-call-api-pop/API/app.js:14]: remove this? the tenant

File: 3-Authorization-II/1-call-api/API/package-lock.json [hunk 12]
```diff
@@ -0,0 +1,2033 @@
+{
+  "name": "ms-identity-react-c3s1",
```
Comment (kalyankrishna1) [inline 3-Authorization-II/1-call-api/API/package-lock.json:2]: >": [](http://example.com/codeflow?start=7&length=2)

delete this file?

**Comment by kalyankrishna1** [comment]
>   "dependencies": {

delete this file?

---
Refers to: 3-Authorization-II/1-call-api/SPA/package-lock.json:6 in 3e8d6a7. [](commit_id = 3e8d6a7d8c3a8f054648fc291b1fd33a202924dc, deletion_comment = False)

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+
+## Contents
+
+| File/folder                         | Description                                                |
+|-------------------------------------|------------------------------------------------------------|
+| `SPA/src/authConfig.js`             | Authentication parameters for SPA project reside here.     |
+| `SPA/src/index.js`                  | MSAL React is initialized here.                            |
+| `API/authConfig.json`               | Authentication parameters for API project reside here.     |
+| `API/app.js`                        | Application entry. passport-azure-ad is initialized here.  |
+| `API/utils/validateToken.js`        | Contains utility methods for validating PoP tokens.        |
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:40]: >validateToken [](http://example.com/codeflow?start=13&length=13)

also call out the is the a file where the key wrap is executed, fetch.js?

File: 6-AdvancedScenarios/2-call-api-pop/README.md [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    cd SPA
+    npm start
+```
+
+In a separate console window, execute the following commands:
+
+```console
+    cd ..
+    cd API
+    node run
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:195]: >node run [](http://example.com/codeflow?start=4&length=8)

npm start

File: 6-AdvancedScenarios/2-call-api-pop/AppCreationScripts/sample.json [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    "AADApps": [
+      {
+        "Id": "service",
+        "Name": "msal-node-api",
+        "Kind": "WebApi",
+        "Audience": "AzureADMyOrg",
+        "HomePage": "https://localhost:5000/api/todolist"
+      },
+      {
+        "Id": "client",
```
Comment (kalyankrishna1) [inline 6-AdvancedScenarios/2-call-api-pop/AppCreationScripts/sample.json:19]: >client [](http://example.com/codeflow?start=15&length=6)

add manual step for apptype=spa

**Review by kalyankrishna1 — APPROVED** [review]
:shipit:

File: 6-AdvancedScenarios/2-call-api-pop/README.md (hunk 13 above)
Comment (derisen) [reply 6-AdvancedScenarios/2-call-api-pop/README.md:40]: added

File: 6-AdvancedScenarios/2-call-api-pop/API/app.js (hunk 11 above)
Comment (derisen) [reply 6-AdvancedScenarios/2-call-api-pop/API/app.js:14]: done

File: 6-AdvancedScenarios/2-call-api-pop/README.md (hunk 3 above)
Comment (idg-sam) [inline 6-AdvancedScenarios/2-call-api-pop/README.md:24]:
```suggestion
1. The React SPA client uses **MSAL React** to sign-in and obtain a JWT access token for the Node.js/Express.js protected resource web API from **Azure AD**.
```

Comment (derisen) [reply 6-AdvancedScenarios/2-call-api-pop/README.md]: done!

**Pull request merged**